### PR TITLE
[sosreport] exit status propagation is broken

### LINKED ISSUE #672: [sosreport] exit status propagation is broken
David Kutalek pointed out:

```
# sosreport -o fooble

sosreport (version 3.2)

a non-existing plugin (fooble) was specified in the command line
# echo $?
0
```

The use of `SystemExit` in `SoSReport._exit()` interacts badly with exception catching blocks at the tail end of `SoSReport.execute()`.

### PR DESCRIPTION
**Pull request by bhavikbhavsar**
fixes #672 
Signed-off-by: Bhavik Bhavsar 9.bhavik@gmail.com

### CODE REVIEW

File: sos/sosreport.py [hunk 1]
```diff
@@ -1609,7 +1609,7 @@ def execute(self):
                     rmtree(self.tmpdir)
             except:
                 raise
-
+            raise to_be_rasied
```
Comment (Akasurde) [inline sos/sosreport.py]: There is spelling mistake, it should be 'to_be_raised' instead of 'to_be_rasied'

**Comment by battlemidget** [comment]
This also doesn't follow our contributor guidelines

**Comment by Amitgb14** [comment]
There two commit, you want to force push to first commit.

**Comment by bhavikbhavsar** [comment]
Thanks @Amitgb14, now there is a single commit. Do let me know if there are any further suggestions.

**Comment by bmr-cymru** [comment]
With the current branch I'm seeing:

```
# sosreport -v --batch
[...]
[plugin:block] collecting output of 'parted -s /dev/dm-6 unit s print'
^CTraceback (most recent call last):
  File "/usr/sbin/sosreport", line 25, in <module>
    main(sys.argv[1:])
  File "/usr/lib/python2.7/site-packages/sos/sosreport.py", line 1619, in main
    sos.execute()
  File "/usr/lib/python2.7/site-packages/sos/sosreport.py", line 1612, in execute
    raise to_be_raised
KeyboardInterrupt
[root@hex tmp]# echo $?
1
```

So the exit status is propagating correctly now but we don't really want to dump a backtrace on an unsuspecting user.

**Comment by bmr-cymru** [comment]
(with `--debug` this traps to pdb - I am in two minds right now whether that is useful or annoying).

**Comment by bmr-cymru** [comment]
@bhavikbhavsar any chance to work on this? 3.3 will be closing very soon.

**Comment by bhavikbhavsar** [comment]
@bmr-cymru yes, I will try to see what would help not to get us stuck in pdb, or if you have any suggestions.

**Comment by bmr-cymru** [comment]
Ping. Any update to this PR? If not, then since there were only minor bugfixes remaining, I will pick it up in the next few weeks and commit it.

**Comment by bhavikbhavsar** [comment]
@bmr-cymru I am a bit busy you can go ahead and fix this.

**Comment by bmr-cymru** [comment]
The changes in `SoSReport.execute()` have two problems:

 * We need to separate `OSError` (an error) from `SystemExit` and `KeyboardInterrupt` (not errors): they cannot share the same `except` block (unless we start poking into the type of the exception, but there's no need for that).

 * It is not correct to re-raise the exception in `execute()`: there is no higher level exception handler in `main()` so these exceptions propagate all the way to the interpreter and end up dumped on the user's console. Instead we should simply exit with an appropriate error code.

I've added a patch on top of this branch that implements those changes.

**Pull request closed without merging**